File: backtest/metrics.py
"""
Performance metrics calculation for backtests.
"""
import logging
from dataclasses import dataclass
from typing import Optional

import numpy as np
import pandas as pd

logger = logging.getLogger(__name__)


@dataclass
class PerformanceMetrics:
    """Comprehensive performance metrics."""
    # Returns
    total_return: float = 0.0
    total_return_pct: float = 0.0
    cagr: float = 0.0
    annualized_return: float = 0.0

    # Risk metrics
    volatility: float = 0.0
    sharpe_ratio: float = 0.0
    sortino_ratio: float = 0.0
    max_drawdown: float = 0.0
    max_drawdown_duration_days: int = 0
    calmar_ratio: float = 0.0

    # Trade statistics
    total_trades: int = 0
    winning_trades: int = 0
    losing_trades: int = 0
    win_rate: float = 0.0
    avg_win: float = 0.0
    avg_loss: float = 0.0
    profit_factor: float = 0.0
    expectancy: float = 0.0
    avg_trade_duration_days: float = 0.0

    # Additional metrics
    best_trade: float = 0.0
    worst_trade: float = 0.0
    avg_trade: float = 0.0
    trading_days: int = 0
    exposure_pct: float = 0.0

    def to_dict(self) -> dict:
        """Convert to dictionary."""
        return {
            "total_return": round(self.total_return, 2),
            "total_return_pct": round(self.total_return_pct, 2),
            "cagr": round(self.cagr, 2),
            "annualized_return": round(self.annualized_return, 2),
            "volatility": round(self.volatility, 2),
            "sharpe_ratio": round(self.sharpe_ratio, 3),
            "sortino_ratio": round(self.sortino_ratio, 3),
            "max_drawdown": round(self.max_drawdown, 2),
            "max_drawdown_duration_days": self.max_drawdown_duration_days,
            "calmar_ratio": round(self.calmar_ratio, 3),
            "total_trades": self.total_trades,
            "winning_trades": self.winning_trades,
            "losing_trades": self.losing_trades,
            "win_rate": round(self.win_rate, 2),
            "avg_win": round(self.avg_win, 2),
            "avg_loss": round(self.avg_loss, 2),
            "profit_factor": round(self.profit_factor, 3),
            "expectancy": round(self.expectancy, 2),
            "avg_trade_duration_days": round(self.avg_trade_duration_days, 1),
            "best_trade": round(self.best_trade, 2),
            "worst_trade": round(self.worst_trade, 2),
            "avg_trade": round(self.avg_trade, 2),
            "trading_days": self.trading_days,
            "exposure_pct": round(self.exposure_pct, 2),
        }

    def format_summary(self) -> str:
        """Format as human-readable summary."""
        return (
            f"📈 **Performance Summary**\n"
            f"Total Return: {self.total_return_pct:+.2f}%\n"
            f"CAGR: {self.cagr:+.2f}%\n"
            f"Sharpe Ratio: {self.sharpe_ratio:.2f}\n"
            f"Max Drawdown: {self.max_drawdown:.2f}%\n\n"
            f"📊 **Trade Statistics**\n"
            f"Total Trades: {self.total_trades}\n"
            f"Win Rate: {self.win_rate:.1f}%\n"
            f"Profit Factor: {self.profit_factor:.2f}\n"
            f"Avg Trade: ${self.avg_trade:.2f}"
        )


class MetricsCalculator:
    """Calculate performance metrics from backtest results."""

    TRADING_DAYS_PER_YEAR = 252
    RISK_FREE_RATE = 0.05  # 5% annual risk-free rate

    def __init__(self, risk_free_rate: float = 0.05):
        """
        Initialize metrics calculator.

        Args:
            risk_free_rate: Annual risk-free rate for Sharpe calculation
        """
        self.risk_free_rate = risk_free_rate

    def calculate(
        self,
        equity_curve: pd.Series,
        trades: list[dict],
        initial_capital: float,
    ) -> PerformanceMetrics:
        """
        Calculate all performance metrics.

        Args:
            equity_curve: Daily equity values (indexed by date)
            trades: List of trade dictionaries
            initial_capital: Starting capital

        Returns:
            PerformanceMetrics object
        """
        metrics = PerformanceMetrics()

        if len(equity_curve) < 2:
            return metrics

        # Basic returns
        final_equity = equity_curve.iloc[-1]
        metrics.total_return = final_equity - initial_capital
        metrics.total_return_pct = (metrics.total_return / initial_capital) * 100

        # Calculate daily returns
        returns = equity_curve.pct_change().dropna()
        metrics.trading_days = len(returns)

        # CAGR
        years = metrics.trading_days / self.TRADING_DAYS_PER_YEAR
        if years > 0 and final_equity > 0 and initial_capital > 0:
            metrics.cagr = ((final_equity / initial_capital) ** (1 / years) - 1) * 100
            metrics.annualized_return = metrics.cagr

        # Volatility (annualized)
        if len(returns) > 1:
            daily_vol = returns.std()
            metrics.volatility = daily_vol * np.sqrt(self.TRADING_DAYS_PER_YEAR) * 100

        # Sharpe Ratio
        if metrics.volatility > 0:
            excess_return = metrics.cagr - (self.risk_free_rate * 100)
            metrics.sharpe_ratio = excess_return / metrics.volatility

        # Sortino Ratio (only downside volatility)
        negative_returns = returns[returns < 0]
        if len(negative_returns) > 0:
            downside_vol = negative_returns.std() * np.sqrt(self.TRADING_DAYS_PER_YEAR) * 100
            if downside_vol > 0:
                excess_return = metrics.cagr - (self.risk_free_rate * 100)
                metrics.sortino_ratio = excess_return / downside_vol

        # Maximum Drawdown
        dd_info = self._calculate_drawdown(equity_curve)
        metrics.max_drawdown = dd_info["max_drawdown"]
        metrics.max_drawdown_duration_days = dd_info["max_duration"]

        # Calmar Ratio
        if abs(metrics.max_drawdown) > 0:
            metrics.calmar_ratio = metrics.cagr / abs(metrics.max_drawdown)

        # Trade statistics
        if trades:
            self._calculate_trade_stats(metrics, trades)

        # Exposure
        in_market_days = sum(1 for t in trades for _ in range(t.get("holding_days", 1) or 1))
        if metrics.trading_days > 0:
            metrics.exposure_pct = (in_market_days / metrics.trading_days) * 100

        return metrics

    def _calculate_drawdown(self, equity_curve: pd.Series) -> dict:
        """Calculate drawdown metrics."""
        rolling_max = equity_curve.expanding().max()
        drawdowns = (equity_curve - rolling_max) / rolling_max * 100

        max_dd = drawdowns.min()

        # Calculate max drawdown duration
        is_underwater = drawdowns < 0
        duration = 0
        max_duration = 0
        for underwater in is_underwater:
            if underwater:
                duration += 1
                max_duration = max(max_duration, duration)
            else:
                duration = 0

        return {
            "max_drawdown": max_dd,
            "max_duration": max_duration,
            "drawdowns": drawdowns,
        }

    def _calculate_trade_stats(
        self,
        metrics: PerformanceMetrics,
        trades: list[dict],
    ) -> None:
        """Calculate trade statistics."""
        pnls = [t.get("pnl", 0) for t in trades if "pnl" in t]

        if not pnls:
            return

        metrics.total_trades = len(pnls)
        metrics.winning_trades = sum(1 for p in pnls if p > 0)
        metrics.losing_trades = sum(1 for p in pnls if p <= 0)

        if metrics.total_trades > 0:
            metrics.win_rate = (metrics.winning_trades / metrics.total_trades) * 100

        wins = [p for p in pnls if p > 0]
        losses = [p for p in pnls if p < 0]

        if wins:
            metrics.avg_win = np.mean(wins)
            metrics.best_trade = max(wins)
        if losses:
            metrics.avg_loss = np.mean(losses)
            metrics.worst_trade = min(losses)

        # Profit Factor
        total_wins = sum(wins) if wins else 0
        total_losses = abs(sum(losses)) if losses else 0
        if total_losses > 0:
            metrics.profit_factor = total_wins / total_losses

        # Expectancy
        if metrics.total_trades > 0:
            metrics.avg_trade = np.mean(pnls)
            metrics.expectancy = (
                (metrics.win_rate / 100 * metrics.avg_win) +
                ((1 - metrics.win_rate / 100) * metrics.avg_loss)
            )

        # Average holding period
        holding_days = [t.get("holding_days", 1) for t in trades if t.get("holding_days")]
        if holding_days:
            metrics.avg_trade_duration_days = np.mean(holding_days)

    def get_drawdown_series(self, equity_curve: pd.Series) -> pd.Series:
        """Get drawdown series for plotting."""
        rolling_max = equity_curve.expanding().max()
        return (equity_curve - rolling_max) / rolling_max * 100

File: rust/backtest-engine/src/metrics.rs
use chrono::{DateTime, Utc};
use common::{PerformanceMetrics, Trade};

const TRADING_DAYS_PER_YEAR: f64 = 252.0;
const RISK_FREE_RATE: f64 = 0.05; // 5% annual risk-free rate

/// Calculate performance metrics from equity curve and trades
pub struct MetricsCalculator;

impl MetricsCalculator {
    /// Calculate all performance metrics
    pub fn calculate(
        equity_curve: &[(DateTime<Utc>, f64)],
        trades: &[Trade],
        initial_capital: f64,
    ) -> PerformanceMetrics {
        if equity_curve.is_empty() {
            return PerformanceMetrics::default();
        }

        let final_equity = equity_curve.last().map(|(_, e)| *e).unwrap_or(initial_capital);
        let total_return = final_equity - initial_capital;
        let total_return_pct = (total_return / initial_capital) * 100.0;

        // Calculate daily returns
        let daily_returns = Self::calculate_daily_returns(equity_curve);

        // Calculate metrics
        let volatility = Self::calculate_volatility(&daily_returns);
        let sharpe_ratio = Self::calculate_sharpe_ratio(&daily_returns, volatility);
        let sortino_ratio = Self::calculate_sortino_ratio(&daily_returns);
        let (max_drawdown, max_dd_duration) = Self::calculate_max_drawdown(equity_curve);

        // CAGR
        let days = equity_curve.len() as f64;
        let years = days / TRADING_DAYS_PER_YEAR;
        let cagr = if years > 0.0 && final_equity > 0.0 && initial_capital > 0.0 {
            ((final_equity / initial_capital).powf(1.0 / years) - 1.0) * 100.0
        } else {
            0.0
        };

        // Calmar ratio
        let calmar_ratio = if max_drawdown != 0.0 {
            cagr / max_drawdown.abs()
        } else {
            0.0
        };

        // Trade statistics
        let (trade_stats, best_trade, worst_trade) = Self::calculate_trade_stats(trades);

        // Exposure percentage
        let exposure_pct = Self::calculate_exposure(equity_curve, trades);

        PerformanceMetrics {
            total_return,
            total_return_pct,
            cagr,
            volatility,
            sharpe_ratio,
            sortino_ratio,
            max_drawdown,
            max_drawdown_duration_days: max_dd_duration,
            calmar_ratio,
            total_trades: trades.len() as u32,
            winning_trades: trade_stats.winning,
            losing_trades: trade_stats.losing,
            win_rate: trade_stats.win_rate,
            avg_win: trade_stats.avg_win,
            avg_loss: trade_stats.avg_loss,
            profit_factor: trade_stats.profit_factor,
            expectancy: trade_stats.expectancy,
            avg_trade_duration_days: trade_stats.avg_duration,
            best_trade,
            worst_trade,
            exposure_pct,
        }
    }

    /// Calculate daily returns from equity curve
    fn calculate_daily_returns(equity_curve: &[(DateTime<Utc>, f64)]) -> Vec<f64> {
        if equity_curve.len() < 2 {
            return vec![];
        }

        equity_curve
            .windows(2)
            .map(|w| {
                let prev = w[0].1;
                let curr = w[1].1;
                if prev != 0.0 {
                    (curr - prev) / prev
                } else {
                    0.0
                }
            })
            .collect()
    }

    /// Calculate annualized volatility
    fn calculate_volatility(daily_returns: &[f64]) -> f64 {
        if daily_returns.is_empty() {
            return 0.0;
        }

        let n = daily_returns.len() as f64;
        let mean: f64 = daily_returns.iter().sum::<f64>() / n;
        let variance: f64 = daily_returns.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / n;

        variance.sqrt() * TRADING_DAYS_PER_YEAR.sqrt() * 100.0
    }

    /// Calculate Sharpe ratio
    fn calculate_sharpe_ratio(daily_returns: &[f64], volatility: f64) -> f64 {
        if daily_returns.is_empty() || volatility == 0.0 {
            return 0.0;
        }

        let n = daily_returns.len() as f64;
        let mean_daily_return = daily_returns.iter().sum::<f64>() / n;
        let annualized_return = mean_daily_return * TRADING_DAYS_PER_YEAR * 100.0;

        (annualized_return - RISK_FREE_RATE * 100.0) / volatility
    }

    /// Calculate Sortino ratio (uses only downside deviation)
    fn calculate_sortino_ratio(daily_returns: &[f64]) -> f64 {
        if daily_returns.is_empty() {
            return 0.0;
        }

        let n = daily_returns.len() as f64;
        let mean_return = daily_returns.iter().sum::<f64>() / n;
        let daily_risk_free = RISK_FREE_RATE / TRADING_DAYS_PER_YEAR;

        // Calculate downside deviation (only negative returns relative to target)
        let downside_returns: Vec<f64> = daily_returns
            .iter()
            .filter(|&&r| r < daily_risk_free)
            .map(|&r| (r - daily_risk_free).powi(2))
            .collect();

        if downside_returns.is_empty() {
            return f64::INFINITY;
        }

        let downside_variance: f64 = downside_returns.iter().sum::<f64>() / n;
        let downside_deviation = downside_variance.sqrt() * TRADING_DAYS_PER_YEAR.sqrt();

        if downside_deviation == 0.0 {
            return 0.0;
        }

        let annualized_return = mean_return * TRADING_DAYS_PER_YEAR;
        (annualized_return - RISK_FREE_RATE) / downside_deviation
    }

    /// Calculate maximum drawdown and duration
    fn calculate_max_drawdown(equity_curve: &[(DateTime<Utc>, f64)]) -> (f64, i64) {
        if equity_curve.is_empty() {
            return (0.0, 0);
        }

        let mut max_equity = equity_curve[0].1;
        let mut max_drawdown = 0.0;
        let mut max_dd_start = 0;
        let mut max_dd_duration = 0i64;
        let mut current_dd_start = 0;

        for (i, (_, equity)) in equity_curve.iter().enumerate() {
            if *equity > max_equity {
                max_equity = *equity;
                current_dd_start = i;
            }

            let drawdown = (max_equity - equity) / max_equity * 100.0;
            if drawdown > max_drawdown {
                max_drawdown = drawdown;
                max_dd_start = current_dd_start;
                max_dd_duration = (i - max_dd_start) as i64;
            }
        }

        (max_drawdown, max_dd_duration)
    }

    /// Calculate drawdown curve
    pub fn calculate_drawdown_curve(
        equity_curve: &[(DateTime<Utc>, f64)],
    ) -> Vec<(DateTime<Utc>, f64)> {
        if equity_curve.is_empty() {
            return vec![];
        }

        let mut max_equity = equity_curve[0].1;
        equity_curve
            .iter()
            .map(|(ts, equity)| {
                if *equity > max_equity {
                    max_equity = *equity;
                }
                let drawdown = if max_equity > 0.0 {
                    (max_equity - equity) / max_equity * 100.0
                } else {
                    0.0
                };
                (*ts, drawdown)
            })
            .collect()
    }

    /// Calculate trade statistics
    fn calculate_trade_stats(trades: &[Trade]) -> (TradeStats, f64, f64) {
        if trades.is_empty() {
            return (TradeStats::default(), 0.0, 0.0);
        }

        let mut winning = 0u32;
        let mut losing = 0u32;
        let mut total_wins = 0.0;
        let mut total_losses = 0.0;
        let mut total_duration = 0i64;
        let mut best = f64::MIN;
        let mut worst = f64::MAX;

        for trade in trades {
            if trade.pnl > 0.0 {
                winning += 1;
                total_wins += trade.pnl;
            } else if trade.pnl < 0.0 {
                losing += 1;
                total_losses += trade.pnl.abs();
            }

            total_duration += trade.holding_days;
            best = best.max(trade.pnl);
            worst = worst.min(trade.pnl);
        }

        let n = trades.len() as f64;
        let win_rate = if n > 0.0 {
            (winning as f64 / n) * 100.0
        } else {
            0.0
        };

        let avg_win = if winning > 0 {
            total_wins / winning as f64
        } else {
            0.0
        };

        let avg_loss = if losing > 0 {
            total_losses / losing as f64
        } else {
            0.0
        };

        let profit_factor = if total_losses > 0.0 {
            total_wins / total_losses
        } else if total_wins > 0.0 {
            f64::INFINITY
        } else {
            0.0
        };

        let expectancy =
            (win_rate / 100.0 * avg_win) - ((1.0 - win_rate / 100.0) * avg_loss);

        let avg_duration = total_duration as f64 / n;

        (
            TradeStats {
                winning,
                losing,
                win_rate,
                avg_win,
                avg_loss,
                profit_factor,
                expectancy,
                avg_duration,
            },
            if best == f64::MIN { 0.0 } else { best },
            if worst == f64::MAX { 0.0 } else { worst },
        )
    }

    /// Calculate exposure percentage
    fn calculate_exposure(
        equity_curve: &[(DateTime<Utc>, f64)],
        trades: &[Trade],
    ) -> f64 {
        if equity_curve.is_empty() || trades.is_empty() {
            return 0.0;
        }

        let total_days = equity_curve.len() as f64;
        let invested_days: i64 = trades.iter().map(|t| t.holding_days.max(1)).sum();

        (invested_days as f64 / total_days * 100.0).min(100.0)
    }
}

#[derive(Debug, Default)]
struct TradeStats {
    winning: u32,
    losing: u32,
    win_rate: f64,
    avg_win: f64,
    avg_loss: f64,
    profit_factor: f64,
    expectancy: f64,
    avg_duration: f64,
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;

    fn make_equity_curve(values: &[f64]) -> Vec<(DateTime<Utc>, f64)> {
        values
            .iter()
            .enumerate()
            .map(|(i, &v)| {
                (
                    Utc.with_ymd_and_hms(2024, 1, 1 + i as u32, 12, 0, 0).unwrap(),
                    v,
                )
            })
            .collect()
    }

    #[test]
    fn test_basic_metrics() {
        let equity = make_equity_curve(&[10000.0, 10100.0, 10200.0, 10300.0, 10400.0]);
        let trades = vec![];

        let metrics = MetricsCalculator::calculate(&equity, &trades, 10000.0);

        assert_eq!(metrics.total_return, 400.0);
        assert_eq!(metrics.total_return_pct, 4.0);
    }

    #[test]
    fn test_max_drawdown() {
        let equity = make_equity_curve(&[10000.0, 11000.0, 9000.0, 9500.0, 10500.0]);
        let (max_dd, _) = MetricsCalculator::calculate_max_drawdown(&equity);

        // Peak was 11000, trough was 9000 = 18.18% drawdown
        assert!((max_dd - 18.18).abs() < 0.1);
    }

    #[test]
    fn test_sharpe_ratio_positive() {
        // Consistently positive returns should give positive Sharpe
        let equity = make_equity_curve(&[
            10000.0, 10100.0, 10200.0, 10300.0, 10400.0, 10500.0, 10600.0,
        ]);
        let metrics = MetricsCalculator::calculate(&equity, &[], 10000.0);

        assert!(metrics.sharpe_ratio > 0.0);
    }

    #[test]
    fn test_drawdown_curve() {
        let equity = make_equity_curve(&[10000.0, 11000.0, 10000.0, 9000.0]);
        let dd_curve = MetricsCalculator::calculate_drawdown_curve(&equity);

        assert_eq!(dd_curve.len(), 4);
        assert_eq!(dd_curve[0].1, 0.0); // No drawdown at start
        assert_eq!(dd_curve[1].1, 0.0); // New high, no drawdown
        assert!((dd_curve[2].1 - 9.09).abs() < 0.1); // 10000/11000 = ~9.09%
        assert!((dd_curve[3].1 - 18.18).abs() < 0.1); // 9000/11000 = ~18.18%
    }
}
